osu.Game.Rulesets.Taiko/Objects/ISkinnableSwell.cs

```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Taiko.Objects.Drawables;
+using osu.Game.Skinning;
+
+namespace osu.Game.Rulesets.Taiko.Objects
+{
+    public interface ISkinnableSwell
+    {
+        void AnimateSwellProgress(DrawableTaikoHitObject<Swell> swell, int numHits, SkinnableDrawable mainPiece);
+
+        void AnimateSwellCompletion(ArmedState state, SkinnableDrawable mainPiece);
+
+        void AnimateSwellStart(DrawableTaikoHitObject<Swell> swell, SkinnableDrawable mainPiece);
```

Comment:
I'm not sure why I didn't complain about this before, but the fact that the swell body is animating the swell piece is really bothering me here. It's a weird dependency that doesn't strictly need to exist.

Proposed direction:

- Rename `ISkinnableSwell` -> `ISkinnableSwellPart`
- Remove the `mainPiece` argument from all of these
- Make all swell main pieces implement `ISkinnableSwellPart` and move their correspondent animations there
- In `DrawableSwell`, collect all `ISkinnableSwellPart` into an `IEnumerable` and every time these methods need to be called, iterate through all parts and call these methods for them

This is probably something that I'd be looking a second reviewer's opinion on, if I can get that.

Reply:
I would avoid the concept of accessing via interface altogether with a set of events or bindables exposed from `DrawableSwell` instead, that are accessed by these pieces via DI. Sort of similar to the skinning flow of the spinner and generally follows the direction skinnable components take (DI'ing components and listening to bindables/events).

Reply:
> generally follows the direction skinnable components take

I think you may be generalising too much here. That's not generally true, other rulesets do similar things. Compare: `IHitExplosion`, `IAnimatableJudgement`, `ICatchComboCounter`.

If `ApplyCustomUpdateState` suffices (I'm not sure it does), then sure, if it can work. But I don't know.

Reply:
By "direction of skinnable components",. I mean the new direction that started from skinnable target support (e.g. skinnable HUD components). But my point is, you should be able to avoid the interface and enumeration completely by exposing events and subscribing to them from the pieces.